packages/envd/internal/services/filesystem/utils.go

```diff
@@ -15,3 +22,45 @@ func getEntryType(entry osEntry) rpc.FileType {
 		return rpc.FileType_FILE_TYPE_FILE
 	}
 }
+
+// getFileOwnership returns the owner and group names for a file.
+// If the lookup fails, it returns the numeric UID and GID as strings.
+func getFileOwnership(fileInfo os.FileInfo) (owner, group string) {
+	sys, ok := fileInfo.Sys().(*syscall.Stat_t)
+	if !ok {
+		return "", ""
+	}
+
+	// Look up username
+	if u, err := user.LookupId(fmt.Sprintf("%d", sys.Uid)); err == nil {
+		owner = u.Username
+	} else {
+		owner = fmt.Sprintf("%d", sys.Uid)
+	}
+
+	// Look up group name
+	if g, err := user.LookupGroupId(fmt.Sprintf("%d", sys.Gid)); err == nil {
+		group = g.Name
+	} else {
+		group = fmt.Sprintf("%d", sys.Gid)
+	}
+
+	return owner, group
+}
+
+func EntryInfoFromFileInfo(fileInfo os.FileInfo, path string) *rpc.EntryInfo {
```

Comment:
could you change it to internal? it isn't used elsewhere
```suggestion
func entryInfoFromFileInfo(fileInfo os.FileInfo, path string) *rpc.EntryInfo {
```